rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,345 @@
+implicit remote communication. For the remainder of the discussion, we focus
+on `exec` only.
+
+### Local Execution
+
+Given the new `exec` command, we can start processing data in situ. For example,
+we can directly compute insights over telemetry:
+
+```bash
+vast exec 'from - |
+           read zeek |
+           summarize sum(orig_bytes), sum(resp_bytes) group-by id.orig_h, id.resp_h |
+           sort |
+           head -n 10 |
+           write json |
+           to -'
+```
+
+This pipeline answers the question "who are the top talkers in my network" by
+summing originator and responder bytes per host pair. The dataflow happens as
+follows:
+
+1. Load data via stdin (`from`)
+2. Push data into a parser (`read`)
+3. Aggregate data (`summarize`)
+4. Sort the aggregate (`sort`)
+5. Only consider the top 10 (`head`)
+6. Push data into a printer (`write`)
+7. Write data to stdout (`to`)
+
+Local pipeline execution opens the door for a lot of new modes of operations,
+because a VAST server node is no longer required.
+
+**This makes VAST a swiss army knife for security data.**
+
+There are of course convenience defaults we can use, like assuming `from -` and
+`to -` being the default. VAST can then easily used as a conversion utility
+between different types of security data:
+
+```bash
+vast exec 'read zeek | write json'
+# Obviates the need for:
+vast exec 'convert zeek json'
+```
+
+### Case Study: Matcher
+
+The `matcher` plugin is an analyzer plugin that operates on the input stream by
+matching an opaque state object against. We call this opaque state object the
+"matcher."
+
+#### Local execution
+
+Let's assume that the following new pipeline operators:
+
+- `build<Arrow, Matcher>`: consumes a data stream to produce a matcher
+- `match<Arrow, Arrow>`: produces sightings for the active matchers
+
+Now consider the use case of constructing a matcher from CSV data:
+
+```bash
+vast exec 'from file:///tmp/blacklist.csv |
+           read csv |
+           build --extract=net.src.ip' > matcher.flatbuf
+```
+
+After having constructed a matcher, we can now use it in a pipeline for
+matching:
+
+```bash
+vast exec 'from kafka -t /zeek/conn |
+           read zeek |
+           match --state matcher.flatbuf --on id.orig_h,id.resp_h
+```
+
+The `match` operator exposes the structured data to a specific matcher instance,
+reshaping according to the positional arguments that are a list of extractors.
+The pipeline writer is responsible for reshaping the data so that the matcher
+can make sense of it.
+
+A successful "match" wraps the corresponding event into a sighting envelope
+record.
+
+## Implementation
+
+The following aspects require attention during the implementation:
+
+- **Asynchronous Execution**: pipelines fundamentally support asynchronous
+  execution by sequentializing dataflow within an operator, and parallelizing it
+  across operators. This maps well to the actor model architecture of VAST.
+  Operators naturally map to actors.
```

Comment:
I don't think operators naturally map to actors. I propose removing that last sentence only, or thoroughly explaining why you think that is the case.

Reply:
Didn't I explain this? VAST's architecture is based on the actor model. Hence it's a natural way to think about them conceptually, given they represent asynchronous dataflow.

Reply:
I agree that the actor model and the actor system underlying our model function well as building blocks for the execution model of pipelines. But that does not mean that there is a mapping from pipeline to actor. Operators as described in the proposal map onto some underlying physical representation, and how they are executed is a yet to be determined implementation detail of the execution model of pipelines. Mapping them onto CAF's actors is just one possible implementation besides CAF's streams, flows, resumables and actions, and even a single threaded implementation is likely feasible.

How users compose pipelines should not dictate how they are being executed, as the pipelines from this proposal only describe the sequence of logical operators that describe what the user wants to do, but not the sequence of physical operators that determine how we actually get to that result.

Reply:
Agreed. I rephrased this to not imply that actors are the main vehicle.

Please resolve if you agree that this is good enough.

Reply:
That works for me. Thanks!